File: stack/monotone_stack/0316_remove_duplicate_letters.js
/*
Given a string s, remove duplicate letters so that every letter appears once 
and only once. You must make sure your result is the smallest in 
lexicographical order among all possible results.
*/

// https://leetcode.com/problems/smallest-subsequence-of-distinct-characters/description/

/* 
get the sequence of letter in s, and construct a monotone stack, 
1) if char > stack[stack.length - 1], push it into stack;
2) if char < stack[stack.length - 1], and we have more than 1 char left, 
   pop until char > stack[stack.length - 1] or stack is empty, then push
3) if stack includes char, skip it, and count[char] - 1 
*/
// time complexity: O(n)
// space complexity: O(n)
var removeDuplicateLetters = function (s) {
  let count = new Map();
  for (let char of s) {
    count.set(char, count.get(char) + 1 || 1);
  }
  let stack = [];
  for (let char of s) {
    if (stack.length === 0) {
      stack.push(char);
    } else if (stack.includes(char)) {
      count.set(char, count.get(char) - 1);
      continue;
    } else if (char < stack[stack.length - 1]) {
      while (
        count.get(stack[stack.length - 1]) > 1 &&
        char < stack[stack.length - 1] &&
        stack.length > 0
      ) {
        let c = stack.pop();
        count.set(c, count.get(c) - 1);
      }
      stack.push(char);
    } else {
      stack.push(char);
    }
  }
  return stack.join("");
};

File: stack/monotonic_stack/0316_remove_duplicate_letters.py
# Given a string s, remove duplicate letters so that every letter appears once 
# and only once. You must make sure your result is the smallest in 
# lexicographical order among all possible results.

# https://leetcode.com/problems/remove-duplicate-letters/
# same problem
# LC 1081 https://leetcode.com/problems/smallest-subsequence-of-distinct-characters/

from collections import Counter
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        left = Counter(s)  # 统计每个字母的出现次数
        ans = []
        in_ans = set()
        for char in s:
            left[char] -= 1
            if char in in_ans:  # ans 中不能有重复字母
                continue
            # (设 x = ans[-1]) 如果 char < x，且右边还有 x，那么可以把 x 去掉，因为后面可以重新把 x 加到 ans 中
            while ans and char < ans[-1] and left[ans[-1]]:
                in_ans.remove(ans.pop())  # 标记 x 不在 ans 中
            ans.append(char)  # 把 char 加到 ans 的末尾
            in_ans.add(char)  # 标记 char 在 ans 中
        return ''.join(ans)
